targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F439xI/TARGET_MODULE_UBLOX_ODIN_W2/sdk/ublox-odin-w2-drivers/OdinWiFiInterface.cpp

```diff
@@ -0,0 +1,1854 @@
+
+    if(status != cbSTATUS_OK) {
+        struct odin_wifi_msg_s* msg = _msg_pool->alloc();
+        MBED_ASSERT(msg != NULL);
+
+        msg->type = ODIN_WIFI_MSG_USER_SCAN_RSP;
+        msg->data.user_scan_rsp.ap_list = 0;
+        msg->data.user_scan_rsp.found_aps = NSAPI_ERROR_DEVICE_ERROR;
+
+        osStatus ok = _out_queue.put(msg);
+        MBED_ASSERT(ok == osOK);
+    }
+}
+
+void OdinWiFiInterface::handle_user_connect_timeout()
+{
+    if((_state_sta == S_STA_WAIT_CONNECT) && (_sta.timeout_ms > 0)) {
+        int elapsed_time = _timer.read_ms();
+
+        if(elapsed_time > _sta.timeout_ms) {
+            if(_debug) {
+                printf("TIMEOUT: %d ms, ACTUAL %d ms\n\r", _sta.timeout_ms, elapsed_time);
+            }
+            _timer.stop();
+
+            _state_sta = entry_connect_fail_wait_disconnect();
+        }
+    }
+}
+
+void OdinWiFiInterface::handle_user_ap_start(user_ap_start_s *user_ap_start)
+{
+    MBED_ASSERT(user_ap_start != 0);
+
+    nsapi_error_t   error_code = NSAPI_ERROR_OK;
+
+    bool channel_ok = is_valid_AP_channel(user_ap_start->channel);
+
+    if((_state_sta == S_STA_IDLE) && (_state_ap == S_AP_IDLE) && channel_ok) {
+        // No STA or AP activity in progress
+
+        error_code = wlan_ap_start(
+                user_ap_start->ssid,
+                user_ap_start->passwd,
+                user_ap_start->security,
+                user_ap_start->channel,
+                user_ap_start->beacon_interval);
+    }
+    else {
+        // Parallel STA or AP activity is not supported
+        error_code = NSAPI_ERROR_UNSUPPORTED;
+    }
+
+    if(error_code == NSAPI_ERROR_OK) {
+        _state_ap = entry_ap_wait_start();
+    }
+    else
+    {
+        send_user_response_msg(ODIN_WIFI_MSG_USER_AP_START, error_code);
+    }
+}
+
+void OdinWiFiInterface::handle_user_ap_stop()
+{
+    switch(_state_ap) {
+    case S_AP_STARTED:
+        _state_ap = entry_ap_wait_stop();
+        break;
+
+    default:
+        send_user_response_msg(ODIN_WIFI_MSG_USER_AP_STOP, NSAPI_ERROR_DEVICE_ERROR);
+        break;
+    }
+}
+
+void OdinWiFiInterface::handle_wlan_status_started(wlan_status_started_s *start)
+{
+    if (_debug) {
+        printf("cbWLAN_STATUS_STARTED\n\r");
+    }
+
+    MBED_ASSERT(start != 0);
+
+    switch(_state) {
+    case S_WAIT_START:
+        sprintf(_mac_addr_str,
+            "%02x:%02x:%02x:%02x:%02x:%02x",
+            start->info.macAddress[0],
+            start->info.macAddress[1],
+            start->info.macAddress[2],
+            start->info.macAddress[3],
+            start->info.macAddress[4],
+            start->info.macAddress[5]);
+
+        if(!_wlan_initialized) {
+            //Initialize network stack interface without activating it
+        }
+
+        if (!_interface) {
+            nsapi_error_t error_code = _stack.add_ethernet_interface(_emac, true, &_interface);
+            if (error_code != NSAPI_ERROR_OK) {
+                _interface = NULL;
+            }
```

Comment:
Like in EMACInterface::connect(), there needs to be call here to "_interface->attach(_connection_status_cb);" in else branch to add the callback to interface. Otherwise callbacks attached between the construct and connect() do not work.

Reply:
is #6562 issue due to this?

Reply:
I think that #7310 could be caused by this.